foreshadow/core/fs_logging.py

```diff
@@ -0,0 +1,264 @@
+
+    try:
+        if _logger is not None:
+            return _logger
+
+        # Get scoped Foreshadow logger.
+        _my_logger = _logging.getLogger("foreshadow")
+
+        _interactive = False
+        try:
+            # Only defined in interactive shells.
+            _interactive = True if _sys.ps1 else _interactive
+        except AttributeError:
+            # check python -i
+            _interactive = _sys.flags.interactive
+
+        if _interactive:
+            _my_logger.setLevel(LEVELS["info"])
+        else:
+            _my_logger.setLevel(LEVELS["warning"])
+        _stream_target = _sys.stderr
+
+        # Add Stream Handler based on if interactive or not.
+        _handler = _logging.StreamHandler(_stream_target)
+        _handler.setFormatter(_logging.Formatter(_logging.BASIC_FORMAT, None))
+        _my_logger.addHandler(_handler)
+
+        _logger = _my_logger
+        return _logger
+
+    finally:
+        _logger_lock.release()
+
+
+def _get_log_fn(level):
+    """Get the logging function for a particular method from this module.
+
+    Used internally in this script as the module will not yet have the
+    methods set.
+
+    Args:
+        level: The corresponding log function level. E.g. 'debug'
+
+    Returns:
+        The log function callable.
+
+    Raises:
+        ValueError: if logging level does not have a logging function.
+
+    """
+    log_fn = getattr(_this_module, level, None)
+    if log_fn is None:
+        raise ValueError("level: '{}' is not a valid logging level.")
+    return log_fn
+
+
+class SyncWrite(object):
+    """Single-instance object to handle buffered writing to file."""
+
+    # TODO turn this into synchronous Queue based writing.
+    def __init__(
+        self, buffer_size=100, outfile="gui_data.txt", overwrite=True
+    ):
+        super(SyncWrite, self).__init__()
+        self.threads = []
+        self._buffer_size = buffer_size
+        self.outfile = outfile
+        self.first_write = overwrite
+        self.buffer = []
+        self.writing = False
+
+    @property
+    def buffer_size(self):
+        """How long to wait before writing.
+
+        Returns:
+            buffer size as integer
+
+        """
+        return self._buffer_size
+
+    @buffer_size.setter
+    def buffer_size(self, value):
+        """Set buffer size to value and ensure it makes sense.
+
+        Args:
+            value: integer to set buffer_size to.
+
+        """
+        if value < 1:
+            log_fn = _get_log_fn("warning")
+            log_fn("trying to set buffer size < 1. Automatically setting to 1")
+            value = 1
+        if value < len(self.buffer):
+            log_fn = _get_log_fn("warning")
+            log_fn(
+                "trying to set buffer < current buffer length. "
+                "Automatically writing buffer to file before setting."
+            )
+            self.write()
+        self._buffer_size = value
+
+    def write(self):
```

Comment:
Would it be worth looking into [RotatingFileHandler](http://www.blog.pythonlibrary.org/2014/02/11/python-how-to-create-rotating-logs/) and [docs](https://docs.python.org/3.8/library/logging.handlers.html#logging.handlers.RotatingFileHandler)

Reply:
Uh, this is for the GUI, it was more just a placeholder until we decide what we want the GUI to look like. Personally I would use some form of synchronous write system not attached to logging.